storage/jellyfish_merkle/src/lib.rs

```diff
@@ -3,4 +3,54 @@
 
 #![allow(dead_code)]
 
+#[cfg(test)]
+mod mock_tree_store;
 mod node_type;
+mod tree_cache;
+
+use crypto::HashValue;
+use failure::prelude::*;
+use node_type::{Node, NodeKey};
+use std::collections::{HashMap, HashSet};
+use types::transaction::Version;
+
+/// The hardcoded maximum height of a [`JellyfishMerkleTree`] in nibbles.
+const ROOT_NIBBLE_HEIGHT: usize = HashValue::LENGTH * 2;
+
+/// `TreeReader` defines the interface between [`JellyfishMerkleTree`] and underlying storage
+/// holding nodes.
+pub trait TreeReader {
+    /// Get node given a node key.
+    fn get_node(&self, node_key: &NodeKey) -> Result<Node>;
+}
+
+/// Node batch that will be written into db atomically with other batches.
+pub type NodeBatch = HashMap<NodeKey, Node>;
+/// RetireLog batch that will be written into db atomically with other batches.
+pub type RetireLogBatch = HashSet<RetiredStateRecord>;
+
+/// Indicates a record becomes outdated on `version_retired`.
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct RetiredStateRecord {
```

Comment:
A blob is a node as well now, so this can be `RetiredNode`, if you like it more. (Although it's me who named it `record`, I never thought it's ideal.)

Reply:
renamed it to `StaleNodeIndex`